app/models/package_manager/base/bulk_version_updater.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module PackageManager
+  class Base
+    # This class is responsible for taking a project and a list of ApiVersions (aka raw versions mapped
+    # to a common set off attributes), and create Versions for the Project in the database, in a single
+    # query. Because we're using +upsert_all()+, Versions that already exist will just be updated. 
+    # This single query avoids all the N+1s involved in a regular batch insert of Versions that we did in the past.
+    class BulkVersionUpdater
+      def initialize(db_project:, api_versions:, repository_source_name:)
+        @db_project = db_project
+        @api_versions = api_versions
+        @repository_source_name = repository_source_name
+      end
+
+      def run!
+        # create synthetic versions without saving them, so we can get their attributes
+        attrs = @api_versions
+          .map { |api_version| Version.new(api_version.to_version_model_attributes.merge(project: @db_project)) }
+          .each do |v|
+            # this value will get merged w/existing in the upsert_all query (see table tests in spec)
+            v.repository_sources = @repository_source_name if @repository_source_name
+            # from Version#before_save
+            v.update_spdx_expression
+            # upsert_all doesn't do validation, so ensure they're valid here.
+            v.validate!
+          end
+          .map { |v| v.attributes.without("id", "created_at", "updated_at") }
+
+        existing_version_ids = @db_project.versions.all.pluck(:id)
+
+        # TODO: we could do this in batches if performance does not scale well with # of versions.
+        Version.upsert_all(
+          attrs,
+          # handles merging any existing repository_sources with new repository_source:
+          #   Prev       New       Result
+          #   ["Main"]  ["Maven"]  ["Main", "Maven"]
+          #   [nil]     ["Maven"]  ["Maven"]
+          #   ["Main"]  [nil]      ["Main"]
+          #   [nil]     [nil]      nil
+          on_duplicate: Arel.sql(%!
+            repository_sources = (CASE
+            WHEN (versions.repository_sources IS NULL AND EXCLUDED.repository_sources IS NULL)
+              THEN NULL
+            WHEN (versions.repository_sources @> EXCLUDED.repository_sources)
+              THEN versions.repository_sources
+            ELSE
+              (COALESCE(versions.repository_sources, '[]'::jsonb) || COALESCE(EXCLUDED.repository_sources, '[]'::jsonb))
+            END)
```

Comment:
this took a while to get right, but it should simulate the logic we had before of appending `repository_sources`. See the spec for a table test of examples